### PR TITLE
Sim api

### PR DESCRIPTION
**Pull request by BGluth**
This is the initial version of the runner.

I wrote a decent amount up in a notebook (`Using the Runner.ipynb`), but the general idea is to have something to make it easy to tie data generation with different types of simulations.

Currently it's hooked up to all the data generators (blob, random, and from file) as well as Maharsh's stuff. The goal is for this to be able very easy to expand for when we add new types of parameters/data generation/simulations.

### CODE REVIEW

File: runner.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+        print('Generating "{}" data...'.format(data_gen_type))
+        generated_data = data_gen_run_info.run_func(self._params)
+
+        print('Running the "{}" simulation...'.format(sim_type))
+        sim_run_info.run_func(self._params, generated_data)
+
+        print("Finished!")
+
+    def set_num_samples(self, num_samples):
```
Comment (crankycoder) [inline runner.py]: Let's get rid of most of these setters.  It doesn't make sense that someone would call one or two of the setters without calling all of the setters required to make the Runner get into a valid overall state.

I think it makes sense to just pass in a JSON blob of configuration to the Runner in the constructor, and then call the state validation code in `_sim_has_required_params_for_given_run_func`

File: runner.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        return self._set_param(P_KEY_BATCH_SIZE, batch_size)
+
+    def set_num_epochs(self, num_epochs):
+        return self._set_param(P_KEY_NUM_EPOCHS, num_epochs)
+
+    def _set_param(self, p_key, val):
+        self._params[p_key] = val
+        return self
+
+    def _sim_has_required_params_for_given_run_func(
```
Comment (crankycoder) [inline runner.py]: Let's call this from the Runner constructor so that we can't ever have a Runner in an invalid startup state.

File: runner.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
+import runner_run_funcs
+
+SIM_TYPE_FED_LEARNING = "fed_learning"
```
Comment (crankycoder) [inline runner.py]: These constants don't have a global semantic meaning.  Let's scope these to be class attributes of `Runner`

File: runner.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+P_KEY_NUM_LABELS = "num_labels"
+P_KEY_NUM_FEATURES = "num_features"
+P_KEY_NUM_USERS = "num_users"
+
+P_KEY_NUM_ROUNDS = "num_rounds"
+P_KEY_BATCH_SIZE = "batch_size"
+P_KEY_NUM_EPOCHS = "num_epochs"
+
+
+class RunFuncAndReqParams:
```
Comment (crankycoder) [inline runner.py]: I think we can safely drop this class - it doesn't 'carry it's own weight'.   A code smell for a class that can be removed is that you've got internal state, but no methods attached to the class.

I would replace this with either a simple 2-tuple, or possibly a dict.

File: runner.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            P_KEY_NUM_EPOCHS,
+            P_KEY_NUM_SAMPLES,
+            P_KEY_NUM_FEATURES,
+            P_KEY_NUM_USERS,
+        },
+    ),
+    SIM_TYPE_FED_AVG_WITH_DP: RunFuncAndReqParams(
+        runner_run_funcs.run_fed_avg_with_dp, {P_KEY_NUM_LABELS, P_KEY_NUM_FEATURES}
+    ),
+    DATA_GEN_TYPE_DATA_FROM_FILE: RunFuncAndReqParams(
```
Comment (crankycoder) [inline runner.py]: Can we split run_func_ltable into two separate lists: 

  * (model, required params) 
  * (data generator, labels)

**Review by crankycoder — COMMENTED** [review]
I think if we simplify the structure a little we can get this into shape so we have something that can easily be imported into a thin server module.   I don't want to dwell too much on nits - let's optimize for readability at this point.

File: runner.py (hunk 4 above)
Comment (BGluth) [reply runner.py]: Yeah I agree. A tuple might make more sense for two fields when it's self contained within a module.

File: runner.py (hunk 5 above)
Comment (BGluth) [reply runner.py]: So... I just need some clarification here.

From our meeting, I know we want the way the simulator gets data to feel more "natural" in terms of how this algo. would run in the real world. But I'm still not totally sure what this would look like. At this point should we still even support data generation right in the runner, or only allow reading in from a file?

I think in the example above  `(data generator, labels)` `data generator` is the run function to generate/read-in the data. Is this right? And would `labels` just be the number of labels in the data?

File: runner.py (hunk 1 above)
Comment (BGluth) [reply runner.py]: Agreed. This is probably nicer to work with.

Should I worry about providing an API to create the JSON?

File: runner.py (hunk 2 above)
Comment (BGluth) [reply runner.py]: Sure.

File: runner.py (hunk 3 above)
Comment (BGluth) [reply runner.py]: Yeah this makes sense once we get rid of the other classes in the module.

File: runner.py (hunk 1 above)
Comment (crankycoder) [reply runner.py]: No, I don't think an API to create the JSON is necessary.

File: runner.py (hunk 5 above)
Comment (crankycoder) [reply runner.py]: re `(data generator, labels)` - I think so - it's hard to get an idea for the ergonomics of what the API is going to be like without iterating a bunch.  Let's just split the monolithic list for now so that we have smaller parts that are easier to test and instrument so that we can iterate 'cheaply' in terms of effort while we figure out how we want to make this work.

Comment (BGluth) [reply runner.py]: Ok I'll give this a try

File: runner.py (hunk 3 above)
Comment (BGluth) [reply runner.py]: I'm wondering now if it actually does make more sense for these to be loose in `runner.py` (or maybe even in a file of their own).

Currently there are multiple files that are dependent on these keys, `runner.py` itself and `runner_run_funcs.py`, and also now any code that's going to be constructing the json. It feels kind of weird to me for these to be defined as statics on `Runner` while in `runner_run_funcs` we don't care at all about the `Runner` class besides these static keys that live on it.

This is kind of nit picky, so IDK. I'd be interested to know your thoughts on this.

Comment (BGluth) [reply runner.py]: It's also kind of weird since both `runner.py` and `runner_run_funcs.py` are deps of eachother.

Comment (BGluth) [reply runner.py]: There is a dependency issue happening when importing `runner.py`.

`runner` imports `runner_run_funcs`, which needs the static key variables from `Runner` (class), but `runner` hasn't been fully imported yet so it doesn't see it as existing.

I can't see a nice way for these to be static without this issue. There might be (if there is, just post a comment here and I'll make the changes it), but for now I'm going to put the keys loose in their own file.

**Comment by BGluth** [comment]
Ok I think I'm ready for another look. @crankycoder 

The only thing that I didn't fully yet implement is moving the data generation ltable stuff to a format like you suggested `(data generator, labels)`. But I did split the sim/datagen run functions into two separate tables.

Also let me know what you think about that dependency issue between `runner.py` and `runner_run_functions.py`.

Also I accidentally rebased onto master when I needed to do a squash, so I ended up including a bunch of commits in the rebase that were not needed.

File: runner.py (hunk 3 above)
Comment (crankycoder) [reply runner.py]: Can we just cheat and just those two modules together for now?

File: runner.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import json
+
+
+class Runner:
+
+    run_func_ltable, data_gen_func_ltable = runner_run_funcs.get_run_funcs()
+
+    def __init__(self, param_json, sim_type, data_gen_type):
+        self._params = json.loads(param_json)
+        self._run(sim_type, data_gen_type)
```
Comment (crankycoder) [inline runner.py]: You should probably rename `_run` to `_init` or just inline the code into the `__init__` method.

File: runner.py (hunk 2 above)
Comment (crankycoder) [reply runner.py]: `_sim_has_required_params_for_given_run_func` should probably throw an exception if invalid parameters have been passed in.

File: runner.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            sim_prereq_params, sim_type
+        ):
+            return
+
+        if not self._sim_has_required_params_for_given_run_func(
+            gen_prereq_params, data_gen_type
+        ):
+            return
+
+        print('Generating "{}" data...'.format(data_gen_type))
```
Comment (crankycoder) [inline runner.py]: The `_run` method seems to have two concerns.  The parts above are initialization related, and the code below is calling `gen_run_func` and `sim_run_func` are actually 'running' the simulation.  Let's split this up so that we have everything initialized at Runner construction time and then executed when run() is invoked.

**Review by crankycoder — COMMENTED** [review]
I've got a couple minor nits to split up the `_run` method into an `_init()` (or inline it) and `run()` method, but this looks pretty good to me.

File: runner.py (hunk 3 above)
Comment (BGluth) [reply runner.py]: Haha sure. :)

**Comment by BGluth** [comment]
Ok @crankycoder, the ball is in your court again.

I made all the requested changes, but I'll be dead honest, merging all three files together is very, very, ugly IMHO.

**Review by crankycoder — APPROVED** [review]
This looks good @BGluth! 

Yes - I agree.  The structure isn't optimal yet, but we're trying to do three things here really:

1. implement the federated learning algorithm
2. add a differential privacy version
3. engineer this so that we can deploy to production

I think what you've got now gives us a rough outline of how we can start to think about running simulations against the different implementations.  

What we've got now is a good place to start iterating to get this into a automatically testable state which will get us #3 easier.
